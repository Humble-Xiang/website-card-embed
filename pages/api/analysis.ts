import { NextApiRequest, NextApiResponse } from 'next';
import AnalysisFactory from '@humble.xiang/website-description-analysis';

const analysis = async (_: NextApiRequest, res: NextApiResponse) => {
  res.status(200).json((await AnalysisFactory.create(_.query.url as string)).analysis());
};

export default analysis;
