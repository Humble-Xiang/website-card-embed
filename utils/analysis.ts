import http from 'http';
import https from 'https';
import * as cheerio from 'cheerio';
import { CheerioAPI } from 'cheerio';

export class AnalysisFactory {
  static async create(url: string): Promise<Analysis> {
    if (url.indexOf('juejin') !== -1) return await new JueJinAnalysisImpl(url).init();
    return await new AnalysisImpl(url).init();
  }
}

export abstract class Analysis {
  private readonly _url: string;
  protected _html!: string;
  private $!: CheerioAPI;

  constructor(url: string) {
    this._url = url;
  }

  async init(): Promise<Analysis> {
    this._html = await curl(this._url);
    this.$ = cheerio.load(this._html);
    return this;
  }

  analysis(): AnalysisData {
    return {
      url: this.url,
      title: this.title,
      description: this.description,
      image: this.image,
      favicon: this.favicon,
    };
  }

  get url(): string {
    return this._url;
  }

  get title(): string {
    if (this.$('h1[class*=title]').length === 1) return this.$('h1[class*=title]').text().trim();
    if (this.$('h1[class*=Title]').length === 1) return this.$('h1[class*=Title]').text().trim();
    if (this.$('.title').length === 1) return this.$('.title').text().trim();
    if (this.$('h1>a').length === 1) return this.$('h1>a').text().trim();
    if (this.$('header h1').length === 1) return this.$('header h1').text().trim();
    if (this.$('h1').length === 1) return this.$('h1').text().trim();
    return this.$('title').text().trim();
  }

  get description(): string | undefined {
    if (this.$('meta[property=og:description]').length === 1)
      return this.$('meta[property=og:description]').attr('content');
    return this.$('meta[name=description]').attr('content');
  }

  get image(): string | undefined {
    let img;
    if (this.$('meta[name=og:image]').length === 1) {
      img = this.$('meta[name=og:image]').attr('content');
    } else if (this.$('div[class*=content] p img').length > 0) {
      img = this.$('div[class*=content] p img').first().attr('src');
    } else if (this.$('div[class*=Content] p img').length > 0) {
      img = this.$('div[class*=content] p img').first().attr('src');
    } else {
      img = this.$('meta[property=og:image]').attr('content');
    }
    return img ? `https://images.weserv.nl/?url=${img}` : undefined;
  }

  get favicon(): string {
    return `https://t3.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&fallback_opts=TYPE,SIZE,URL&url=${this._url}&size=16`;
  }
}

class AnalysisImpl extends Analysis {}

class JueJinAnalysisImpl extends Analysis {
  private readonly titleRegExp: RegExp = /(?<=headline":")[^"]*/;

  get title(): string {
    const match = this.titleRegExp.exec(this._html);
    if (match) return match[0];
    return '';
  }
}

export type AnalysisData = {
  readonly url: string;
  readonly title: string;
  readonly description: string | undefined;
  readonly image: string | undefined;
  readonly favicon: string;
};

async function curl(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    (url.startsWith('https') ? https : http)
      .get(url, res => {
        let data = '';
        res.on('data', function (chunk) {
          data += chunk;
        });
        res.on('end', () => {
          resolve(data);
        });
      })
      .on('error', error => {
        reject(error);
      });
  });
}
