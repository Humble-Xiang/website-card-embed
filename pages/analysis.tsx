import { useRouter } from 'next/router';
import useSWR, { Fetcher } from 'swr';
import { AnalysisData } from '../utils/analysis';
import WebsiteCardIntroductionLoading from '@/components/website-card-introduction-loading';
import WebsiteCardIntroduction from '@/components/website-card-introduction';

function AnalysisPage() {
  const url = useRouter().query.url as string;
  const { isLoading, analysisData } = useAnalysisData(url);
  if (isLoading) return WebsiteCardIntroductionLoading();
  if (analysisData) return WebsiteCardIntroduction(analysisData);
  return WebsiteCardIntroductionLoading();
}

const analysisDataFetcher: Fetcher<AnalysisData, string> = url => fetch(url).then(res => res.json());

function useAnalysisData(url: string) {
  if (url) console.info(`useAnalysisData with url:${url}`);
  const { data, error } = useSWR(url ? `/api/analysis?url=${url}` : null, analysisDataFetcher);
  if (error) console.error(error);
  return {
    analysisData: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export default AnalysisPage;
