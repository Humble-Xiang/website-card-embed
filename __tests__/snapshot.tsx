import { render } from '@testing-library/react';
import WebsiteCardIntroduction from '@/components/website-card-introduction';
import { AnalysisFactory } from '@/utils/analysis';

describe('analysis snapshot', () => {
  it('analysis https://zhuanlan.zhihu.com/p/59838643', async () => {
    const analysisData = (await AnalysisFactory.create('https://zhuanlan.zhihu.com/p/59838643')).analysis();
    expect(analysisData).toMatchSnapshot({ image: expect.any(String) });
  });

  it('analysis https://www.cnblogs.com/yichong/p/9234265.html', async () => {
    const analysisData = (await AnalysisFactory.create('https://www.cnblogs.com/yichong/p/9234265.html')).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot();
  });

  it('analysis https://blog.csdn.net/design_Lu/article/details/94870265', async () => {
    const analysisData = (
      await AnalysisFactory.create('https://blog.csdn.net/design_Lu/article/details/94870265')
    ).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot();
  });

  it('analysis https://segmentfault.com/a/1190000020387433', async () => {
    const analysisData = (await AnalysisFactory.create('https://segmentfault.com/a/1190000020387433')).analysis();
    expect(analysisData).toMatchSnapshot({ image: expect.any(String) });
  });

  it('analysis https://www.jianshu.com/p/eec5e34ff0c2', async () => {
    const analysisData = (await AnalysisFactory.create('https://www.jianshu.com/p/eec5e34ff0c2')).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot();
  });

  it('analysis https://stackoverflow.com/questions/43431550/async-await-class-constructor/43433773', async () => {
    const analysisData = (
      await AnalysisFactory.create(
        'https://stackoverflow.com/questions/43431550/async-await-class-constructor/43433773'
      )
    ).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot('stackoverflow');
  });

  it('analysis https://blog.51cto.com/u_15349616/3717558', async () => {
    const analysisData = (await AnalysisFactory.create('https://blog.51cto.com/u_15349616/3717558')).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot();
  });

  it('analysis https://juejin.cn/post/6844904009887645709', async () => {
    const analysisData = (await AnalysisFactory.create('https://juejin.cn/post/6844904009887645709')).analysis();
    const { container } = render(<WebsiteCardIntroduction {...analysisData} />);
    expect(container).toMatchSnapshot();
  });
});
