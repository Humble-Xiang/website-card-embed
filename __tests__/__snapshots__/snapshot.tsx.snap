// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot build WebsiteCardIntroduction 1`] = `
<div>
  <div
    class="css-gmuwbf"
  >
    <div
      class="chakra-linkbox css-gfjtic"
    >
      <div
        class="css-da13x"
      >
        <div
          class="chakra-stack css-whissx"
        >
          <h4
            class="css-zx51qw"
          >
            <a
              class="chakra-linkbox__overlay css-1hnz6hu"
              href="https://humble-blog.vercel.app/"
              target="_blank"
            >
              向之礼的个人博客
            </a>
          </h4>
          <div
            class="css-0"
          >
            <p
              class="chakra-text css-1s2aj5y"
            >
              一个打工人的个人博客小站
            </p>
          </div>
          <div
            class="chakra-stack css-k7bpxb"
          >
            <img
              alt="website card icon"
              class="chakra-image css-f6brzi"
              src="https://t3.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&fallback_opts=TYPE,SIZE,URL&url=https://humble-blog.vercel.app/&size=16"
            />
            <p
              class="chakra-text css-m70ba5"
            >
              https://humble-blog.vercel.app/
            </p>
          </div>
        </div>
        <img
          alt="website card image"
          class="chakra-image css-1nvrwv4"
          src="https://images.weserv.nl/?url=https://humble-blog.vercel.app/images/avatar.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot build WebsiteCardIntroductionLoading 1`] = `
<div>
  <div
    class="css-gmuwbf"
  >
    <div
      class="css-1y8o4jq"
    >
      <div
        class="chakra-stack css-whissx"
      >
        <div
          class="chakra-skeleton css-16lwe9v"
        />
        <div
          class="chakra-skeleton css-1ou8gbg"
        />
        <div
          class="chakra-skeleton css-80kv3h"
        />
      </div>
      <div
        class="chakra-skeleton css-i2ejzm"
      />
    </div>
  </div>
</div>
`;
